rasterio/features.py

```diff
@@ -340,18 +321,29 @@ def format_cast_error(param, dtype):
             else:
                 raise ValueError("Invalid or empty shape cannot be rasterized.")
 
-    if out is not None:
-        if _getnpdtype(out.dtype).name not in valid_dtypes:
-            raise ValueError(format_invalid_dtype('out'))
+    # If neither an out array or dtype were given, we get the output
+    # data type from the shapes values, including the default.
+    if not dtype and valid_shapes:
+        values_arr = np.array(shape_values)
+        dtype = values_arr.dtype.name
+        assert dtype in ("int64", "float64")
```

Comment:
Instead of assert, I recommend raising an exception with an error message.

Reply:
Though, may not even need the assert here.

Reply:
Yeah, not strictly needed. It's only a check against unexpected numpy behavior or `str` or `object` somehow getting into the values.

Reply:
Actually we're not doing much to prevent non-numbers from getting into the values, but I would like to defer work on that. Maybe it could be a part of work on optimizing the marshaling of shapes and values (in Cython).